### ISSUE TITLE
 Possibility to search system tags with a better query system

### ISSUE THREAD

**Issue by Spartachetto** [nextcloud/server#4851]
It would be really useful to have the possibility to search not only for tag A AND tag B but also for tag A AND NOT tag B, or tag A OR tag B

I would value especially tag A AND NOT tag B.
Some use cases:

- All the documents about client X and not about contracts
- All the photos about family and not about private

I was thinking that even a"simple" change in the interface would be a big improvement.
![andnot](https://cloud.githubusercontent.com/assets/12546982/26021372/108398a4-378c-11e7-9b74-cade9d5c020d.png)

My hypothesis of how to make it work would be:

- automatic focus on "Select tags to filter by" as now
- start to search when the first tag in "... filter by" field is selected
- allow to fill "... to exclude" even if "... filter by" is empty, yet do NOT start to search if "... filter by" is empty

This feature should take into account #2143, which is already part of #4699 and scheduled for Nextcloud 13

**Comment by MorrisJobke** [comment]
cc @nextcloud/designers

**Comment by jancborchardt** [comment]
A first change here could be a syntax like:

> labelname not:label-to-exclude

Similar to how other search services do it.

**Comment by Spartachetto** [comment]
If you choose to use syntax, you could also provide something a bit more powerful/complex with parentheses and _OR_ operator.
Maybe the best solution could be an "easy" interface with limited functionality plus possibility to use syntax to have more power.

**Comment by MariusBluem** [comment]
Have a look how GitHub's Issue Tracker does.

- visual filter in "Tags"-menu, which will apply as syntax in the search bar.
- syntax in search bar

**Comment by compumatter** [comment]
I am running into a problem that may highlight a "use case" for you here: https://help.nextcloud.com/t/limit-tags-scope-for-shared-folders/20560

**Comment by skjnldsv** [comment]
Do we want tags in unified search @jancborchardt ?

**Comment by Spartachetto** [comment]
**_Pretty please_**!

On a more serious tone: the unified search seems a really interesting development. 

I do understand the idea that a lot of users do not use tags, yet I think that if Nextcloud offer that functionality it should offer also (as far as possible) tools to use it properly. 

As you are redefining search and vastly improving it, my impression is that not including tags in it would be a missed opportunity.

Thank you for al the improvements you are developing!

**Comment by skjnldsv** [comment]
> Thank you for al the improvements you are developing!

Thanks :)
The issue with tags is that listing all direct tagged ressources isn't really that great on Nextcloud (the files tag view is awful)

So I would be up for this once we have a proper tags view :thinking:

**Comment by Spartachetto** [comment]
Clearly _better tags view_  + _better search tools for tags_ would be terrific!!!

Is there any issue for the _better tags view_? 
I do not remember seeing it, yet if it exists it could be helpful to link it here...

**Comment by arrowd** [comment]
One feature that probably should be easy to implement is to allow searching for tags by any part of the string, not only the beginning.

In my instance, I have a lot of tags like `from:foo`, `from:baz`, etc. Typing `foo` when adding a tag results in 0 matches, so I have to type `from:`, which results in a lot of irrelevant tags.

**Comment by sp33dsk8** [comment]
The default tag filtering appears broken to me,  If I have 3 cards one with x tag one with y tag and one with z tag, and I select to filter on x and y tags, the filter will only search cards by AND logic not by OR logic so I get zero results.  But I see we already have a general idea on how to move forward with this I think.

**Comment by stefan-niedermann** [comment]
@sp33dsk8 you are completely off topic.
This issue is about the global search function of Nextcloid itself and not about the filters in the Deck app.
The issue you are looking for already exists, though: https://github.com/nextcloud/deck/issues/2149

**Comment by KernSani** [comment]
Would like to bring up this topic again. In general tags are really powerful, however how they are currently implemented for files, they are pretty useless (Nextcloud 30.0.1):
* Unified search doesn't support searching for multiple tags (search syntax would probably have to be enhanced for that, ideally with boolean operators
* In the files application only a single tag can be selected (so a tag is similar to a folder) - would suggest a filter possibility (like we have it for document types, but with checkboxes to have the possibility to select multiple tags). E.g. the deck app has such a filter

**Comment by FelixKainz** [comment]
> In general tags are really powerful, however how they are currently implemented for files, they are pretty useless

Yes, please! Tags can be soo powerful. I cannot live without them.
The poor tagging search is my biggest issue with Nextcloud :/

**Comment by mschxyz** [comment]
>   In the files application only a single tag can be selected (so a tag is similar to a folder)

In the older versions of the files app it was possible to search for multiple tags, which was an **amazing** feature.
I'm not entirely sure why it's not possible anymore and how difficult it would be to bring it back.

As an example: We organize quite a few files concerning public events (think something like concerts).
When looking for the print files of posters from all events in 2024, I'd just type `print`, `poster` and `2024`.

Now I can only search for a single tag.
So looking for `print` will give me everything we ever printed, `poster` will give me every poster from every event and `2024` will give me everything from the year 2024, including contracts, photos and whatnot.

**Comment by support-econocom-lu** [comment]
Hello,

I would like to bring this up too.
We have several project where we wanted to use tags to classify documents.

We can't use tags as expected as there is no way to search for multiple tags.
For instance, If we want to find all documents owned by IT regarding HPe servers, we should search for tag IT and HPe and Servers.

In the current state, we can only search for IT or for HPe or for Server tag, so it's unfortunately pretty useless to us.

Regards,

**Comment by Io-Qo** [comment]
I would also underline that this has existed in the past and within my organisation we built our file structure around the possibility to search for multiple tags. Then this has - without any warning - been removed and literally people had to manually sort thousands of files back into separate folders to make them findable again. Making this change was rather careless in my opinion so I would also beg the developers to bring it back and make sure it stays.

I want to summarise the advantages:
- being able to virtually sort files into multiple "folders" with tags when you can't actually sort one file into multiple folders without creating copies
- assigning tags with flow can be automated (!)
- file retention is based on tags
- files access control is based on tags

So many of Nextcloud's inherent features evolve around tags, searching for tag combinations seems only natural and also shouldn't really be a big hurdle to implement.

**Comment by fschiefenhoevel** [comment]
+1 from me for the re-introduction of the capability to search for multiple tags!

**Comment by CodePeasants** [comment]
+1 this feature would be very useful to me.
